Fix upload for empty checksum

kugaevsky opened the pull request:
Fixes #7

chaadow commented:
Hi,
Thank you for your contribution.
I believe checksum is always sent by rails for security measures. Therefore avoiding to send it to the openstack provider as an `etag` could lead to attacks.

Can you provide me with a case in which the checksum is null? a stacktrace for instance, or a minimal rails application in which you replicate the change.

kugaevsky commented:
Please, take a look at `DiskService#upload` method.
https://github.com/rails/rails/blob/master/activestorage/lib/active_storage/service/disk_service.rb#L21

chaadow commented:
That's the `DiskStorage` 😄  . Take a look at one of the "remote" providers. They always send the checksum like this one  From AWS S3: https://github.com/rails/rails/blob/master/activestorage/lib/active_storage/service/s3_service.rb#L22

kugaevsky commented:
@chaadow you right, but there is always `nil` as a default value for checksum. So if you call `upload` method with nil checksum there will be an error.

chaadow commented:
That's in the method definition. but rails always sends the file checksum. take a look at the tests.
If you get a nil checksum, that means you're not uploading the file the right way ( or maybe not passing any 'io' to the `#upload` method.
And the error you get is normal and is expected, since having no valid checksum (or a nil checksum for that matter) means there is an integrity error, hence halting the upload process altogether.

kugaevsky commented:
@chaadow Uploading files runs perfectly even without code in this PR. But when I'm trying to get a variant calling `variant(options)` I'm getting an integrity error. 

I'm not passing any `io` for original file here. Just trying to get already uploaded file with another dimension: #7

chaadow commented:
Yes i know the io is passed automatically by Rails. Can you provide me with your gemfile? Did you add `image_processing`: https://edgeguides.rubyonrails.org/active_storage_overview.html#transforming-images

If so, did you install `mini magick` as well

kugaevsky commented:
Okay, I'm investigated further and found that `variant` calls `upload` without any checksum. It happens here: https://github.com/rails/rails/blob/v5.2.1/activestorage/app/models/active_storage/variant.rb#L130

The checksum is calculated only for the original file https://github.com/rails/rails/blob/v5.2.1/activestorage/app/models/active_storage/blob.rb#L155

It seems that you have to handle the situation with `nil` checksum in `upload` method. Fixed the PR to do so.

kugaevsky commented:
@chaadow fixed conflicts with `master` branch after merging other PRs.

kugaevsky commented:
@chaadow 
Also, all gem tests are failing right now, cause you're calling `upload` method without setting checksum in tests — https://github.com/chaadow/activestorage-openstack/blob/master/test/active_storage/service/open_stack_service_test.rb#L14

If your tests are green then it seems that your OpenStack service doesn't check file integrity when uploading.

Another thing is that dummy application is using `disk` storage in `test` environment. That's wrong as well.

I'll try to fix tests a little bit later and create another PR.

chaadow commented:
@kugaevsky My tests pass, you can upload without setting the checksum.
https://github.com/chaadow/activestorage-openstack/blob/master/test/active_storage/service/open_stack_service_test.rb#L21

Please have a thorough look. My provider does check file integrity. :)
Of course any PR is welcome. thank you again for your effort

kugaevsky commented:
Yes, exactly this test can pass. 8)
Cause you're sending checksum. What about all others?

1. https://github.com/chaadow/activestorage-openstack/blob/master/test/active_storage/service/open_stack_service_test.rb#L14
2. https://github.com/chaadow/activestorage-openstack/blob/master/test/active_storage/service/open_stack_service_test.rb#L69
3. https://github.com/chaadow/activestorage-openstack/blob/master/test/active_storage/service/open_stack_service_test.rb#L151-L153
4. https://github.com/chaadow/activestorage-openstack/blob/master/test/active_storage/service/open_stack_service_test.rb#L171

In all these cases you're calling `upload` without checksum argument.

chaadow commented:
All of these tests are passing. All the test suite for that matter.  Otherwise @d--j would have fixed it in the other PR. Have a look at the PR and the test suite. he just added some new tests for the new methods. All the other tests are passing well.

My guess is your provider does not handle Etags. Hence the tests are failing for you. Could you check your provider's documentation?

Basically, if you send a nil checksum to your provider then it will not check for integrity. However, if you send a checksum, and your provider detects a different one, then it will raise an error.

chaadow commented:
In conclusion, if you don't specify the `Etag` header (or send an empty param), then obviously your provider cannot check for integrity.

kugaevsky commented:
My provider is checking `Etag` for sure . I believe the reason for failing tests is that my provider doesn't ignore empty `Etag` param and tries to check the integrity in this case.

kugaevsky commented:
My point is that we should handle this scenario as well.

chaadow commented:
Your point is fair enough. I understand. Different providers implement the checks differently I guess. Can you then apply your change ( by not sending the Etag if checksum is nil), and re run your tests locally? and let me know if those make you pass the tests?

kugaevsky committed:
Fix upload for empty checksum

- Remove empty `Etag` header for `nil` checksum in request
- Check for the presence of `content-disposition` in unit tests

kugaevsky commented:
Found another issue in my provider's OpenStack implementation. It doesn't send `Content-Disposition` header. Fixed this too.

All tests pass.

chaadow commented:
Alright thank you again for your contribution. Can you rebase on top of master so that there is only 1 commit in the PR.
Afterwards, I will review directly in the commits tab

kugaevsky commented:
@chaadow Done!

Thanks again for your time.

kugaevsky commented:
@chaadow Is everything ok with PR right now?

chaadow commented on lib/active_storage/service/open_stack_service.rb:
```diff
@@ -16,10 +16,9 @@ def initialize(container:, credentials:, connection_options: {})
-          'ETag' => convert_base64digest_to_hexdigest(checksum)
-        }
+        params = { 'Content-Type' => guess_content_type(io) }
+        params['ETag'] = convert_base64digest_to_hexdigest(checksum) if checksum
```
The purpose of this gem is to be compatible with all OpenStack providers. I'm going to approve and merge this PR, but could you **send me** the OpenStack wrapper library you're using. I might submit a PR to rails as to send the checksum with variants as well. ( I don't get why they don't as you've mentioned) 
it's still considered an upload, or maybe i'm missing something

kugaevsky replied:
Do you mean the OpenStack wrapper library from my cloud provider?

chaadow replied:
Yes, but actually don't mind the question @kugaevsky

chaadow reviewed (commented):
Overall, the PR is good to go 👍

chaadow reviewed (approved):
👌

chaadow reviewed (changes requested):
Could you rename your commit to be like this : "Fix upload for empty checksum"

And then put a description like this: 
- Remove empty Etag header for nil checksum in request
- Check for the presence of `content-disposition` in unit tests

kugaevsky commented:
@chaadow done!

kugaevsky commented:
@chaadow 
> I'm going to approve and merge this PR, but could you send me the OpenStack wrapper library you're using.

Finally, my OpenStack provider replied to me. They using Juno with some custom patches.

chaadow commented:
@kugaevsky alright thank you. i'd like to know your opinion about the other subject in the discussion( for which there is a [PR](https://github.com/chaadow/activestorage-openstack/pull/11)
